presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -776,27 +857,103 @@ public Page getNextPage()
                 if (qualifyingSet == null) {
                     qualifyingSet = new QualifyingSet();
                 }
-                qualifyingSet.setRange((int) currentGroupRowCount);
+                qualifyingSet.setRange(0, Math.min(ariaBatchRows, (int) currentGroupRowCount));
                 reader.setQualifyingSets(qualifyingSet, null);
+                if ((currentRowGroup & 0x3) == 0) {
+                    // Decay row size stats every 4 groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                }
                 if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
                     // Reconsider filter order every 4 row groups.
                     reader.maybeReorderFilters();
                 }
+                if (numResults > 0 && reader.mustExtractValues(currentRowGroup == 0)) {
+                    return resultPage();
+                }
+            }
+            if (qualifyingSet.isEmpty()) {
+                qualifyingSet.setRange(qualifyingSet.getEnd(), Math.min(qualifyingSet.getEnd() + ariaBatchRows, (int) currentGroupRowCount));
             }
-            qualifyingSet.clearTruncationPosition();
-            reader.advance();
+            int bytesBeforeAdvance = reader.getResultSizeInBytes();
+            int numResultsBeforeAdvance = reader.getNumResults();
+            // Decimates qualifyingSet by any column-independent
+            // nondeterministic filters.
+            applyNonDeterministicConstantFilters();
+            try {
+                if (!qualifyingSet.isEmpty()) {
+                    reader.advance();
+                }
+            }
+            catch (BatchTooLargeException e) {
```

Comment:
Should we be using an exception for this logic? Exceptions are intended for "exceptional" events, ex. IOException, etc. 
Would it be possible to use logic that doesn't involve exceptions for this, like:
have `reader.advance` return a boolean and if it's false then execute the code in the catch block.

There is a lot of overhead when an exception is thrown (filling in the stacktrace, etc.)